### DISCUSSION TITLE
When someone uses an unroutable room ID ("Failed to join"), suggest using via server derived from the room ID

### DISCUSSION THREAD

**Discussion by MadLittleMods** [element-hq/element-meta#974] [Ideas]
### Problem

When someone uses an unroutable room ID, it currently just presents them with an error modal that looks like following and isn't helpful for knowing what's wrong or how to fix it:

> Failed to join
>
> MatrixError: [404] No known servers (https://matrix-client.matrix.org/_matrix/client/r0/join/!abc%3Agitter.im)

![](https://user-images.githubusercontent.com/558581/217679558-1f3472a3-3d5c-4d2a-82ab-8a4030b0e639.png)

### Potential solution

We could instead suggest a best-guess `via` server derived from the room ID. Yes, this does mean parsing an opaque room ID (*gasps*) to give the guess, but the user still needs to do the correct thing in the end.

This not only helps the user get where they want but also teaches them for next time.

> Failed to join
> 
> The `!JsbthXjlaqosxISjgl:gitter.im` room isn't known to your homeserver (`matrix.org`).
> 
> You can try using a "via" server to join through which is supplied as the 2nd parameter in the `/join` slash command: `/join !JsbthXjlaqosxISjgl:gitter.im gitter.im`
> 
> Or if you're using a matrix.to link, you can try adding a `?via=gitter.im` query parameter to the URL: `https://matrix.to/#/!JsbthXjlaqosxISjgl:gitter.im?via=gitter.im`
>
> Alternatively, if you know a room alias (starts with a `#`) associated with this room, that will always be routable and allow you to join.
>
> Error details (expand/collapse)

![](https://user-images.githubusercontent.com/558581/217683868-e2fc4c85-40d6-4116-b29e-70b7d66536b8.png)

We can even include `matrix.org` as a fallback as that is probably a great server to use as a shot in the dark for things to work out. We want to delight and teach people, not bury them caveats.

### Have you considered any alternatives?

Ideally, this via server assumption could just be automatic but I don't think this would hold up against decision makers with a strong room IDs are opaque strings opinions. And also doesn't teach people for a future where we can't assume.

### Additional context

Related issues:

 - https://github.com/vector-im/element-web/issues/2925
 - https://github.com/vector-im/element-web/issues/17740
 - https://github.com/vector-im/element-web/issues/13247
    - https://github.com/matrix-org/matrix-react-sdk/pull/6048
 - https://github.com/vector-im/element-web/issues/11756 
 - https://github.com/vector-im/element-web/issues/22720
 - https://spec.matrix.org/v1.5/appendices/#routing

 ---
 
 
 <details>
 <summary>Mockup HTML/CSS</summary>
 
 ```html
 <div class="mx_Dialog_content" id="mx_Dialog_content" style="
    line-height: 2.2rem;
">
<p>The <code>!JsbthXjlaqosxISjgl:gitter.im</code> room isn't known to your homeserver (<code>matrix.org</code>).</p>
<p>You can try using a "via" server to join through which is supplied as the 2nd parameter in the /join slash command: <code>/join !JsbthXjlaqosxISjgl:gitter.im gitter.im</code></p>
<p>Or if you're using a matrix.to link, you can try adding a <code>?via=gitter.im</code> query parameter to the URL: <a href="https://matrix.to/#/!JsbthXjlaqosxISjgl:gitter.im?via=gitter.im">https://matrix.to/#/!JsbthXjlaqosxISjgl:gitter.im?via=gitter.im</a></p>
<p>Alternatively, if you know a room alias (starts with a <code>#</code>) associated with this room, that will always be routable and allow you to join.</p>

<details>
<summary>Error details</summary>
<pre><code>...</code></pre>
</details>

</div>
```

```css
#mx_Dialog_content code {
    background-color: #f4f6fa;
    border: 1px solid #e3e8f0;
    border-radius: 4px;
    padding: 2px 4px;
    line-height: 2.2rem;
    margin-right: 2px;
}
```
 
 
 </details>

**Comment by abhijain2003** [comment]
that's a better approach to help end-user. can I work on this?

**Comment by JeanPaulLucien** [comment]
Anyone knows what "404" means. This is not client issue. You are going to offer to user spend his time.
`The !JsbthXjlaqosxISjgl:gitter.im room isn't known to your homeserver (matrix.org).`
Yeah, because homeserver for this room is gitter.im, not matrix.org.

**Comment by abhijain2003** [comment]
and I think most client won't be able to understand the routing / and all, so

**Comment by JeanPaulLucien** [comment]
I checked server gitter.im. It is really lagged. Ping shows "Time out".
![gitter im is lagged](https://user-images.githubusercontent.com/24779886/218045346-875d51d8-42f2-48da-bb7d-34cfdaa9b342.jpg)

I'm as end-user would not spend my time end computing power for lagged servers.

I found Gitter room address displayed as `#...:gitter.im` I joined without issue.

**Comment by t3chguy** [comment]
> Anyone knows what "404" means. This is not client issue. You are going to offer to user spend his time.
> `The !JsbthXjlaqosxISjgl:gitter.im room isn't known to your homeserver (matrix.org).`
> Yeah, because homeserver for this room is gitter.im, not matrix.org.

Rooms don't have a specific homeserver, they're shared amongst all servers which have users in the room.

**Comment by t3chguy** [comment]
> We could instead suggest a best-guess via server derived from the room ID. Yes, this does mean parsing an opaque room ID (gasps) to give the guess, but the user still needs to do the correct thing in the end.

If this is a heuristic we want then it should be implemented in the spec and server rather than every client where it'll be inconsistent

As for defaulting to matrix.org, leaking room IDs of private federations would be considered a privacy issue.

**Comment by JeanPaulLucien** [comment]
Seems room `!JsbthXjlaqosxISjgl:gitter.im` does not exist.

**Reply by JeanPaulLucien** [reply]
It's 0 users room.

**Comment by t3chguy** [comment]
@JeanPaulLucien this isn't a support room for specific rooms.

**Reply by JeanPaulLucien** [reply]
I tried to reproduce at least 5 issues about 404 error code. How many were reproduced? 0.
Nobody confirmed that room exist, it's not 0 users.
